File: lib/tmdb_api/model/media.dart
import 'package:flutter/foundation.dart';
import 'package:tmdb/tmdb_api/model/tv.dart';

import 'media_type.dart';
import 'movie.dart';
import 'person.dart';

class Media {
  bool adult;
  int id;
  MediaType mediaType;
  double popularity;

  Media({
    this.adult,
    @required this.id,
    this.mediaType,
    this.popularity,
  }) : assert(id != null);

  /// Creates a [Media] from a JSON object.
  factory Media.fromJson(Map<String, dynamic> json) {
    if (json == null) return null;

    return Media(
      adult: json['adult'],
      id: json['id'],
      mediaType: MediaType.fromJson(json['media_type']),
      popularity: json['popularity']?.toDouble(),
    );
  }

  factory Media.fromJsonType(Map<String, dynamic> json) {
    if (json == null) return null;

    final mediaType = MediaType.fromJson(json['media_type']) ?? MediaType.all;

    switch (mediaType) {
      case MediaType.movie:
        return Movie.fromJson(json);
      case MediaType.person:
        return Person.fromJson(json);
      case MediaType.tv:
        return Television.fromJson(json);
      default:
        return Media.fromJson(json);
    }
  }

  DateTime date() {
    return null;
  }

  String getTitle() {
    return null;
  }
}

File: react-native/src/tmdb_api/model/Media.ts
import { MediaType } from "./MediaType"
import { Movie, MovieClass } from "./Movie"
import { Person, PersonClass } from "./Person"
import { Television, TelevisionClass } from "./Television"

export interface Media {
 
    adult: boolean
    id: number
    media_type: MediaType
    popularity: number

// factory Media.fromJsonType(Map < String, dynamic > json) {
//     if (json == null) return null

//     let mediaType = media_type ?? MediaType.all

//     switch (mediaType) {
//         case MediaType.movie:
//             return Movie.fromJson(json)
//         case MediaType.person:
//             return Person.fromJson(json)
//         case MediaType.tv:
//             return Television.fromJson(json)
//         default:
//             return Media.fromJson(json)
//     }
// }
}

export abstract class MediaClass {
    static date(media: Media): Date | null {
        let mediaType = media.media_type
        switch (mediaType) {
            case MediaType.movie:
                return MovieClass.date(media as Movie)
            case MediaType.person:
                return PersonClass.date(media as Person)
            case MediaType.tv:
                return TelevisionClass.date(media as Television)
            default:
                return null
        }
    }

    static displayTitle(media: Media): string {
        let mediaType = media.media_type
        switch (mediaType) {
            case MediaType.movie:
                return MovieClass.displayTitle(media as Movie)
            case MediaType.person:
                return PersonClass.displayTitle(media as Person)
            case MediaType.tv:
                return TelevisionClass.displayTitle(media as Television)
            default:
                return ""
        }
    }
}